executor/common_linux.h

```diff
@@ -1485,6 +1485,218 @@ static long syz_io_uring_submit(volatile long a0, volatile long a1, volatile lon
 
 #endif
 
+#if SYZ_EXECUTOR || __NR_syz_btf_id_by_name
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+// Some items in linux/btf.h are relatively new, so we copy them here for
+// backward compatibility.
+#define BTF_MAGIC 0xeB9F
+
+struct btf_header {
+	__u16 magic;
+	__u8 version;
+	__u8 flags;
+	__u32 hdr_len;
+	__u32 type_off;
+	__u32 type_len;
+	__u32 str_off;
+	__u32 str_len;
+};
+
+#define BTF_INFO_KIND(info) (((info) >> 24) & 0x0f)
+#define BTF_INFO_VLEN(info) ((info)&0xffff)
+
+#define BTF_KIND_INT 1
+#define BTF_KIND_ARRAY 3
+#define BTF_KIND_STRUCT 4
+#define BTF_KIND_UNION 5
+#define BTF_KIND_ENUM 6
+#define BTF_KIND_FUNC_PROTO 13
+#define BTF_KIND_VAR 14
+#define BTF_KIND_DATASEC 15
+
+struct btf_type {
+	__u32 name_off;
+	__u32 info;
+	union {
+		__u32 size;
+		__u32 type;
+	};
+};
+
+struct btf_enum {
+	__u32 name_off;
+	__s32 val;
+};
+
+struct btf_array {
+	__u32 type;
+	__u32 index_type;
+	__u32 nelems;
+};
+
+struct btf_member {
+	__u32 name_off;
+	__u32 type;
+	__u32 offset;
+};
+
+struct btf_param {
+	__u32 name_off;
+	__u32 type;
+};
+
+struct btf_var {
+	__u32 linkage;
+};
+
+struct btf_var_secinfo {
+	__u32 type;
+	__u32 offset;
+	__u32 size;
+};
+
+#define BTF_VMLINUX_FIX_ADDRESS ((char*)0x01000000)
+
+// Read out all the content of /sys/kernel/btf/vmlinux to the fixed address
+// buffer and return it. Return NULL if failed.
+static char* read_btf_vmlinux()
+{
+	int fd = open("/sys/kernel/btf/vmlinux", O_RDONLY);
+	if (fd < 0)
+		return NULL;
+
+	struct stat stat;
+	if (fstat(fd, &stat) < 0)
+		return NULL;
+	unsigned long vmlinux_size = stat.st_size;
+	// Read the whole vmlinux file into buf.
+	char* buf;
+	buf = (char*)mmap(BTF_VMLINUX_FIX_ADDRESS, vmlinux_size, PROT_READ | PROT_WRITE,
+			  MAP_SHARED | MAP_ANONYMOUS | MAP_FIXED_NOREPLACE, -1, 0);
+
+	if ((void*)buf == MAP_FAILED) {
+		if (errno == EEXIST) {
+			// The mmap already exists and we will assume this is the one that
+			// was made by the previous call to this function.
+			return BTF_VMLINUX_FIX_ADDRESS;
+		} else {
+			return NULL;
+		}
+	}
+
+	if (buf != BTF_VMLINUX_FIX_ADDRESS) {
+		fail("The address is different even though we issued "
```

Comment:
does fail terminate our generated program?  It seems sort of odd that our error handling here is inconsistent with returning null when mmap fails.  Would it be better to munmap buf and return NULL instead?

Reply:
So this happens only when `mmap` with `MAP_FIXED_NOREPLACE` succeed, but `mmap` returns a different address, which should be impossible unless there is a bug in kernel.

Originally I think it as an assertion only. I agree that it is somewhat inconsistent. I will change the test `buf == MAP_FAILED` to `buf != BTF_VMLINUX_FIX_ADDRESS` then.